silverstripe-terraformers/keys-for-cache
# Bugfix: Set reading mode when reading/writing CacheKeys

Fixes #62
Closes #62

Basically related to the discovery here:
https://github.com/silverstripe/silverstripe-versioned/issues/382

If a user/author/etc is in a `LIVE` reading mode then any `write()` (to any `Versioned` `DataObject`) is essentially a publish (just without the extension hooks being triggered).

The change here is to make sure that we always retrieve and `write()` our `CacheKeys` in the expected stage.

## Manual testing

So, for all of these, I just made my Page and Blocks output the value of `$CacheKey` into the template so that I could see what it was, and then I compared between `?stage=Stage` and `?stage=Live`.

**Test One:**

1. Publish page
2. View page in `?stage=Stage` and note CacheKey values
3. View page in `?stage=Live` and compare values
4. **Result:** As expected, they were the same

**Test Two:**

1. Make a change to the Page and save (not publish)
2. View page in `?stage=Stage` and note CacheKey values
3. View page in `?stage=Live` and compare values
4. **Result:** As expected, the `Stage` hashes had updated, and the `Live` hashes were the same
5. Publish page
6. View page in `?stage=Stage` and note CacheKey values
7. View page in `?stage=Live` and compare values
8. **Result:** As expected, the `Live` hashes match `Stage`

**Test Three:**

1. Delete all `CacheKey` records from all tables
2. View page in `?stage=Stage`
3. **Result:** As expected, there were records in the Draft and Versions tables, but no records in the Live table
2. View page in `?stage=Live`
3. **Result:** As expected, Live table was updated, and there were new "published" records in the Versions table

## Test coverage

So ah, yea, this got really large.

All of our scenarios now run between different stages, and some even perform different actions (`write()` vs `publishRecursive()` between those stages.

You'll see this data provider a lot. The explanation of what each test does, and what is expected, is hopefully described ok.

```php
public function readingModesWithSaveMethods(): array
{
    return [
        // If write() is performed on a model then we would expect the CacheKey to be updated in DRAFT only. Since
        // we are working in the DRAFT stage, we would expect a different value when we fetch that CacheKey again
        'performing write() in DRAFT stage' => [Versioned::DRAFT, 'write', false],
        // If publishRecursive() is performed on a modal, then we expect the same behaviour as above for the DRAFT
        // stage of our CacheKey
        'performing publishRecursive() in DRAFT stage' => [Versioned::DRAFT, 'publishRecursive', false],
        // If write() is performed on a model then we would expect the CacheKey to be updated in DRAFT only. Since
        // we are working in the LIVE stage, we would expect the LIVE value of this CacheKey to be unchanged
        'performing write() in LIVE stage' => [Versioned::LIVE, 'write', true],
        // If publishRecursive() is performed on a modal, then we expect that CacheKey to also be published. As we
        // are working in the LIVE stage, we would now expect a new CacheKey value when it if fetched again
        'performing publishRecursive() in LIVE stage' => [Versioned::LIVE, 'publishRecursive', false],
    ];
}
```
